File: Medium/Number of occurrence/number-of-occurrence.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>

using namespace std;


// } Driver Code Ends
//User function template for C++
class Solution{
public:	
	/* if x is present in arr[] then returns the count
		of occurrences of x, otherwise returns 0. */
	int count(int arr[], int n, int x) {
	    // code here
	   // int lbind = lower_bound(arr,arr+n,x) - arr;
	   // if(lbind == n)
	   // {
	   //     return 0;
	   // }
	   // int ubind = upper_bound(arr,arr+n,x) - arr;
	   // int highind = ubind - 1;
	    
	   // return (highind - lbind + 1);
	   
	   int low = 0;
	   int high = n - 1;
	   int firstind = -1;
	   
	   while(low <= high)
	   {
	       int mid = low + (high - low)/2;
	       
	       if(arr[mid] == x)
	       {
	           firstind = mid;
	           high = mid - 1;
	       }
	       else if(arr[mid] < x)
	       {
	           low = mid + 1;
	       }
	       else
	       {
	           high = mid - 1;
	       }
	   }
	   
	   if(firstind == -1)
	   {
	       return 0;
	   }
	   
	   low = 0;
	   high = n - 1;
	   int lastind = -1;
	   
	   while(low <= high)
	   {
	       int mid = low + (high - low)/2;
	       
	       if(arr[mid] == x)
	       {
	           lastind = mid;
	           low = mid + 1;
	       }
	       else if(arr[mid] > x)
	       {
	           high = mid - 1;
	       }
	       else
	       {
	           low = mid + 1;
	       }
	   }
	   
	   //int lastind = upbound - 1;
	   
	   return (lastind - firstind + 1);
	}
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, x;
        cin >> n >> x;
        int arr[n];
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        auto ans = ob.count(arr, n, x);
        cout << ans << "\n";
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Easy/Number of occurrence/number-of-occurrence.java
class Solution {
    int countFreq(int[] arr, int target) {
        // code here
        int n = arr.length;
        int l = 0;
        int r = n - 1;
        int first = -1;

        while(l <= r)
        {
            int mid = (l + r) >> 1;

            if(arr[mid] == target)
            {
                first = mid;
                r = mid - 1;
            }
            else if(arr[mid] < target)
            {
                l = mid + 1;
            }
            else 
            {
                r = mid - 1;
            }
        }

        if(first == -1) return 0;   
    
        int last = -1;
        l = 0;
        r = n - 1;

        while(l <= r)
        {
            int mid = (l + r) >> 1;
            if(arr[mid] == target)
            {
                last = mid;
                l = mid + 1;
            }
            else if(arr[mid] < target)
            {
                l = mid + 1;
            }
            else
            {
                r = mid - 1;
            }
        }

        return last - first + 1;
    }
    
}
